archive/untar_stream.ts

```diff
@@ -0,0 +1,270 @@
+ *     .readable
+ *     .pipeThrough(new DecompressionStream('gzip'))
+ *     .pipeThrough(new UnTarStream())
+ * ) {
+ *   console.log(entry.pathname)
+ *   entry
+ *     .readable
+ *     ?.pipeTo((await Deno.create(entry.pathname)).writable)
+ * }
+ * ```
+ */
+export class UnTarStream {
+  #lock = false;
+  #readable: ReadableStream<TarStreamEntry>;
+  #writable: WritableStream<Uint8Array>;
+  #gen: AsyncGenerator<Uint8Array>;
+  constructor() {
+    const { readable, writable } = new TransformStream<
+      Uint8Array,
+      Uint8Array
+    >();
+    this.#readable = ReadableStream.from(this.#untar());
+    this.#writable = writable;
+
+    this.#gen = async function* () {
+      let push: Uint8Array | undefined;
+      const buffer: Uint8Array[] = [];
+      for await (let chunk of readable) {
+        if (push) {
+          const concat = new Uint8Array(push.length + chunk.length);
+          concat.set(push);
+          concat.set(chunk, push.length);
+          chunk = concat;
+        }
+
+        for (let i = 512; i <= chunk.length; i += 512) {
+          buffer.push(chunk.slice(i - 512, i));
+        }
+
+        const remainder = -chunk.length % 512;
+        push = remainder ? chunk.slice(remainder) : undefined;
+
+        while (buffer.length > 2) {
+          yield buffer.shift()!;
+        }
+      }
+      if (push) throw new Error("Tarball has an unexpected number of bytes.");
+      if (buffer.length < 2) {
+        throw new Error("Tarball was too small to be valid.");
+      }
+      if (!buffer.every((value) => value.every((x) => x === 0))) {
+        throw new Error("Tarball has invalid ending.");
+      }
+    }();
+  }
+
+  async *#untar(): AsyncGenerator<TarStreamEntry> {
+    const decoder = new TextDecoder();
+    while (true) {
+      while (this.#lock) {
+        await new Promise((a) => setTimeout(a, 0));
+      }
+
+      const { done, value } = await this.#gen.next();
+      if (done) break;
+
+      // Validate Checksum
+      const checksum = value.slice();
+      checksum.set(new Uint8Array(8).fill(32), 148);
+      if (
+        checksum.reduce((x, y) => x + y) !==
+          parseInt(decoder.decode(value.slice(148, 156 - 2)), 8)
+      ) throw new Error("Invalid Tarball. Header failed to pass checksum.");
+
+      // Decode Header
+      let header: OldStyleFormat | PosixUstarFormat = {
+        name: decoder.decode(value.slice(0, 100)).replaceAll("\0", ""),
+        mode: decoder.decode(value.slice(100, 108 - 2)),
+        uid: decoder.decode(value.slice(108, 116 - 2)),
+        gid: decoder.decode(value.slice(116, 124 - 2)),
+        size: parseInt(decoder.decode(value.slice(124, 136)).trimEnd(), 8),
+        mtime: parseInt(decoder.decode(value.slice(136, 148 - 1)), 8),
+        checksum: decoder.decode(value.slice(148, 156 - 2)),
+        typeflag: decoder.decode(value.slice(156, 157)),
+        linkname: decoder.decode(value.slice(157, 257)).replaceAll(
+          "\0",
+          "",
+        ),
```

Comment:
Slicing the string is harder than slicing it as bytes since the string is in UTF-16 and one character over isn't necessarily one byte over which could cause misalignment for the following properties. There is also a lot of `\0` in said string.

Reply:
`subarray` may be faster than `slice` because it doesn't allocate a new `ArrayBuffer`.